cpp/src/parquet/column_writer.cc

```diff
@@ -1301,6 +1301,10 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
     bool single_nullable_element =
         (level_info_.def_level == level_info_.repeated_ancestor_def_level + 1) &&
         leaf_field_nullable;
+    if (!leaf_field_nullable && leaf_array.null_count() != 0) {
```

Comment:
I don't know should we check `single_nullable_element` rather than `leaf_field_nullable`

Reply:
I don't even know what `single_nullable_element` means.

Reply:
As what parquet handle nulls, `leaf_field_nullable` might including nulls in parents?

Reply:
E.g. `List< int notnull> nullable`, we may have the `single_nullable_element == false` but `leaf_field_nullable == true`?

Reply:
`leaf_field_nullable` is computed from `PathBuilder.nullable_in_parent_`, which itself is initialized in e.g.:
https://github.com/apache/arrow/blob/fb8e8122f623f4548b22ece7485c4570d7ece1a6/cpp/src/parquet/arrow/path_internal.cc#L792-L801

Reply:
And for lists:
https://github.com/apache/arrow/blob/fb8e8122f623f4548b22ece7485c4570d7ece1a6/cpp/src/parquet/arrow/path_internal.cc#L725-L741